ArgsParsing/TypeParsers/AnyOrderParser.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using ArgsParsing.Types;
+
+namespace ArgsParsing.TypeParsers
+{
+    /// <summary>
+    /// A parser that parses values using the generic type arguments of any of the derived classes
+    /// of <see cref="AnyOrder"/>, trying out all possible permutations.
+    /// For example parsing <c>AnyOrder&lt;int, string&gt;</c> will succeed for both
+    /// <c>{"123", "foo"}</c> and <c>{"foo", "123"}</c>, but not for <c>{"foo", "bar"}</c>.
+    /// In case of multiple possible permutations, the first matching permutation gets returned.
+    /// </summary>
+    public class AnyOrderParser : BaseArgumentParser<AnyOrder>
+    {
+        private readonly ArgsParser _argsParser;
+
+        public AnyOrderParser(ArgsParser argsParser)
+        {
+            _argsParser = argsParser;
+        }
+
+        // https://stackoverflow.com/a/13022090
+        private static IEnumerable<IList<T>> Permutations<T>(IList<T> values, int fromInd = 0)
+        {
+            if (fromInd + 1 == values.Count)
+            {
+                yield return values;
+            }
+            else
+            {
+                foreach (var v in Permutations(values, fromInd + 1))
```

Comment:
Permutations of a list have factorial growth, obviously. While your stack depth will grow linearly, your execution time will grow factorially.  
By the time recursion becomes a problem, your code will take longer to execute than it will take for the Sun to become a supernova.

Reply:
okay fair Kappa